Add term-description block

aristath opened the pull request:
## Description

This PR adds the `term-description` block. 
Related to https://github.com/WordPress/gutenberg/issues/22724

## How has this been tested?
Tested in an FSE theme by adding the block in an `index.html` template.
* On the front it only appears while in a category/tag/term, and shows the term description
* On the backend it only shows a placeholder.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/native-mobile.md -->

aristath committed:
WIP for term-description block

aristath committed:
load the block.

aristath committed:
better safe than sorry

ntsekouras commented:
Hey Ari 👋 - similar PR opened a few hours ago by @carolinan :https://github.com/WordPress/gutenberg/pull/29609.

Same comment applies here I think: https://github.com/WordPress/gutenberg/pull/29609#pullrequestreview-606086664 and would like your thoughts as well :)

carolinan reviewed (changes requested):
The two pull requests are very similar but since it looks like we are leaning towards this version, please:
- Add text size, line height, text align and color options
- Generate and include the fixtures: https://github.com/WordPress/gutenberg/blob/master/packages/e2e-tests/fixtures/blocks/README.md

aristath committed:
Merge branch 'trunk' into add/term-description

aristath committed:
Add more supports

aristath committed:
add fixtures files

aristath committed:
use padding instead of height so border can adapt depending on font-size

aristath committed:
Change the icon

Use icon from https://github.com/WordPress/gutenberg/pull/27989#issuecomment-755232436

ntsekouras commented on packages/icons/src/library/term-description.js:
```diff
@@ -0,0 +1,20 @@
+	<SVG xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
+		<Path
+			stroke="#1E1E1E"
+			stroke-width="1.5"
```
```suggestion
			strokeWidth="1.5"
```

ntsekouras commented on packages/block-library/src/term-description/index.js:
```diff
@@ -0,0 +1,20 @@
+export { metadata, name };
+
+export const settings = {
+	title: _x( 'Term Description', 'block title' ),
```
We should add a small description here probably.

gziolo replied:
I don't think it matters, because as soon as we move all translations to `block.json` the context is going to be generalized.

ntsekouras replied:
Sorry - I made this confusing :) 

I meant adding a `description` in the block to show in Block Card :)

aristath replied:
Added 👍

ntsekouras commented on packages/block-library/src/term-description/block.json:
```diff
@@ -0,0 +1,20 @@
+	"name": "core/term-description",
+	"category": "design",
+	"attributes": {
+		"textAlign": {
```
You're not using this anywhere 🤔 . You should add the toolbar control `AlignmentToolbar` and handle it in php as well.

aristath replied:
Oh my... how did I miss that? Done 👍

ntsekouras commented on packages/block-library/src/term-description/editor.scss:
```diff
@@ -0,0 +1,11 @@
+.wp-block-term-description__placeholder {
```
I like the approach with the placeholder styles. Just check to have the alignments from attributes and remove the `italic`.

Other changes like `color` etc are shown in preview.

ntsekouras commented on packages/block-library/src/term-description/index.php:
```diff
@@ -0,0 +1,46 @@
+function render_block_core_term_description( $attributes, $content, $block ) {
+
+	$description = false;
+	if ( is_category() ) {
```
`category_description` and `tag_description` are just wrappers of `term_description`, so you can use only that one.

ntsekouras commented on packages/block-library/src/term-description/index.php:
```diff
@@ -0,0 +1,46 @@
+ */
+function render_block_core_term_description( $attributes, $content, $block ) {
+
+	$description = false;
```
I'm wondering if `is_archive` could be enough or correct 😄   🤔 .  `term_description` will not return anything if no term found.

aristath replied:
I used `is_category() || is_tag() || is_tax()` instead. Archives can be for dates and other things, so limiting it to taxonomies seems safer :)

ntsekouras reviewed (commented):
Thanks for working on this Ari! I've left a couple of comments that should be addressed and we can retest again afterwards.

--cc @jasmussen for some design input about the `icon` and placeholder.

jasmussen commented:
> --cc @jasmussen for some design input about the icon and placeholder.

I'd be super happy to help with icon work and tweaks and happy to push directly. I have to work on something else for the remainder of the day, though.

Got any screenshots?

aristath committed:
Merge branch 'trunk' into add/term-description

aristath committed:
Update packages/icons/src/library/term-description.js

Co-authored-by: Nik Tsekouras <ntsekouras@outlook.com>

aristath committed:
address feedback

aristath committed:
Merge remote-tracking branch 'origin/add/term-description' into add/term-description

aristath committed:
Merge branch 'trunk' into add/term-description

aristath committed:
Add description

aristath commented:
Thank you for the feedback @ntsekouras! All notes were addressed. 🎉 

@jasmussen I used the icon from https://github.com/WordPress/gutenberg/pull/27989#issuecomment-755232436

Screenshot with left/center/right aligned blocks, and the icon in use:
![term-description](https://user-images.githubusercontent.com/588688/110630971-ba034080-81ae-11eb-85de-4b4a5aa7377f.png)

The placeholder is similar to the one used for content, but some styles (flex) were removed to make alignments work, and also I removed the `height:100px` that the post-content block has, and replaced it with `padding: 1em 0`. It makes more sense for this block since we can change the font-size and the text can take 2 lines (in which case the 100px we use in the post-content block make it seem broken)

ntsekouras commented on packages/block-library/src/term-description/index.php:
```diff
@@ -0,0 +1,42 @@
+ */
+function render_block_core_term_description( $attributes, $content, $block ) {
+
+	if ( is_category() || is_tag() || is_tax() ) {
```
Would you mind reversing the check to avoid the indentation for the most code?

`if ( ! ..... ) return ''; `

aristath replied:
Done 👍

aristath committed:
reverse condition & logic

jasmussen commented:
Ah cool, seems like an okay icon for now.

About the placeholder text, this is my feeling for the "This is a placeholder for post content" also, we need to do better and be more accurate to the end result. I think there's a longer term opportunity to explore some beautiful prose to pre-fill empty template blocks with, but near term we can still be smarter.

It's essentially this field, correct?

<img width="737" alt="Screenshot 2021-03-10 at 14 12 38" src="https://user-images.githubusercontent.com/1204802/110634835-b53c8d80-81aa-11eb-883d-ddecc3699173.png">

Given how rare these are filled out, I will admit it's hard to find something inspirational here. How about:

"Description for the current term view."

Could also just be "Term description."

Anything to avoid "This is the placeholder for", basically 😅

ntsekouras commented on packages/block-library/src/term-description/index.php:
```diff
@@ -0,0 +1,41 @@
+	if ( ! is_category() && ! is_tag() && ! is_tax() ) {
+		return '';
+	}
+	$classes = 'term-description';
```
`get_block_wrapper_attributes `applies `wp-block-term-description` css class to the block, so no need to init here. That was the intention here, no?

ntsekouras commented on packages/block-library/src/term-description/edit.js:
```diff
@@ -0,0 +1,47 @@
+			</BlockControls>
+			<div { ...blockProps }>
+				<div className="wp-block-term-description__placeholder">
+					<span>
```
Do we need this extra `div` and `span`?

aristath replied:
It's what we have for the post-content block, so for consistency I copy-pasted the implementation.
I agree we most likely don't need it, but I'd rather we change all placeholders at once, keeping them consistent...

aristath committed:
Change placeholder text

ntsekouras commented on packages/block-library/src/term-description/index.php:
```diff
@@ -0,0 +1,41 @@
+	}
+	$wrapper_attributes = get_block_wrapper_attributes( array( 'class' => $classes ) );
+
+	return '<div ' . $wrapper_attributes . '>' . term_description() . '</div>';
```
I'm not 100% sure with this wrapper `div`. Do you think we can avoid it with `term_description` or somehow? Not strong opinions though, just seems extraneous.

aristath replied:
The term-description can contain anyhthing... the field accepts HTML, so it's vital to have a wrapper in order to properly apply styles to it.
I checked into using a filter and it's certainly possible... but it would add another layer of complexity without any real benefit, the end result would be the same, just expressed differently 👍

ntsekouras reviewed (commented):
Took it for another spin. I think it's really close now :)

carolinan commented:
The text size option in the placeholder works better now. I agree that the extra class name can be removed.
It is common for themes to wrap the description inside a div with a custom CSS class, but I think this is a habit inherited from underscores. (_s)

aristath committed:
Simplify

aristath commented:
> Anything to avoid "This is the placeholder for", basically 😅

Changed to `Term description` 👍

aristath commented:
> The text size option in the placeholder works better now. 

Thank you for checking!

> I agree that the extra class name can be removed.

Done.

ntsekouras commented on packages/block-library/src/term-description/editor.scss:
```diff
@@ -0,0 +1,4 @@
+.wp-block-term-description__placeholder {
+	padding: 1em 0;
```
This one (padding) should be removed probably so as to match the end result in front end.

aristath replied:
Other placeholders (like for example the post-content have a `height:100px` for the placeholder, and then add `display:flex` to it. This makes the placeholder text vertically align inside its container and the dashed borders of placeholders don't touch the text, as there is a gap between them.
In this case display:flex and height:100px could not be used. To compensate for that, vertical padding was added to the placeholder 😉 
It's a reasonable compromise, and ideally all placeholders would be refactored in a future PR. What we have now doesn't quite work if we start changing font-sizes etc...

ntsekouras replied:
I'm not sure I understand this 100%. If we just removed the padding we would have other possible styling issues? Because I tried it through dev tools only and seemed identical to front-end. Of course I tested with one theme (tt1-blocks) and you're certainly more experienced in themes :) 

It's a small thing that can be changed in the future if we need to, but want to understand better :)

ntsekouras reviewed (approved):
Thanks for working on this Ari!  🎉 

I have only left a comment about the padding. Other than that it's good to 🚢
